File: Criacionais/Prototype/Exemplo/prototype.py
# Classe Pessoa que será clonada:
class Pessoa:
    def __init__(self, id, nome, idade):
        self.id = id
        self.nome = nome
        self.idade = idade

    # Método clone para realizar uma cópia superficial:
    def clone(self):
        return Pessoa(self.id, self.nome, self.idade)

    def __repr__(self):
        return f"Pessoa(id={self.id}, nome='{self.nome}', idade={self.idade})"


# Classe GerenciaPessoa para gerenciar instâncias de Pessoa:
class GerenciaPessoa:
    def __init__(self):
        self.pessoas = {}

    # Adiciona uma nova pessoa ao dicionário:
    def add_pessoa(self, id, nome, idade):
        pessoa = Pessoa(id, nome, idade)
        self.pessoas[id] = pessoa

    # Retorna uma cópia da pessoa:
    def get_pessoa_by_id(self, id):
        pessoa_original = self.pessoas.get(id)
        if pessoa_original:
            return pessoa_original.clone()
        return None


# Uso:
gerencia = GerenciaPessoa()

gerencia.add_pessoa(1, "João da Silva", 30)
gerencia.add_pessoa(2, "Maria da Silva", 25)
gerencia.add_pessoa(3, "Pedro da Silva", 45)

pessoa_clone = gerencia.get_pessoa_by_id(1)

if pessoa_clone:
    pessoa_clone.nome = "João Clonado"
    pessoa_clone.idade = 99

print("Pessoa original:", gerencia.pessoas[1])
print("Pessoa clonada:", pessoa_clone)

File: Criacionais/Prototype/Exemplo/prototype.js
// Classe Pessoa que será clonada:
class Pessoa {
    constructor(id, nome, idade) {
        this.id = id;
        this.nome = nome;
        this.idade = idade;
    }

    // Metodo clone para realizar uma copia superficial do objeto:
    clone() {
        return new Pessoa(this.id, this.nome, this.idade);
    }
}

// Classe GerenciaPessoa para gerenciar instancias de Pessoa:
class GerenciaPessoa {
    constructor() {
        this.pessoas = {};
    }

    // Add uma nova pessoa ao dicionario de pessoas:
    addPessoa(id, nome, idade) {
        const pessoa = new Pessoa(id, nome, idade);
        this.pessoas[id] = pessoa;
    }

    // Solicita uma pessoa pelo id e retorna uma copia dela:
    getPessoaById(id) {
        const pessoaOriginal = this.pessoas[id];

        if (pessoaOriginal) {
            return pessoaOriginal.clone();
        } else {
            return null;
        }
    }
}

// Criando uma instancia de GerenciaPessoa:
const gerencia = new GerenciaPessoa();

// Adicionando Pessoas:
gerencia.addPessoa(1, "João da Silva", 30);
gerencia.addPessoa(2, "Maria da Silva", 25);
gerencia.addPessoa(3, "Pedro da Silva", 45);

// Clonando a primeira pessoa e modificando o nome da copia
const pessoaClone = gerencia.getPessoaById(1);

if (pessoaClone) {
    pessoaClone.nome = "João Clonado";
    pessoaClone.idade = 99;
}

// Exibindo resultados
console.log("Pessoa original:", gerencia.pessoas[1]);
console.log("Pessoa clonada:", pessoaClone);